### PR TITLE
Refactor Template Engine for Spring Boot 2.x projects

### PR DESCRIPTION
**Pull request by christophe-f**
* Add the TemplateProcessor class
* From the ComponentController, if the Application Type is Spring Boot 2.x, it will use the Template Processor
* Merge from master

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -63,6 +63,7 @@ Including one of the following in the request body:
 {
   "name": "boot-app-2",
   "group": "scotiabank",
+  "groupId": "com.scotiabank",
```
[the remark this answers is not in the document]
Comment (christophe-f) [reply README.md]: I saw that when doing the merge. I thought the group will be the replacement of the old project key which was more for teams. If now it's the same as groupId, I will merge into one. Also which name is best?

File: build.gradle [hunk 2]
```diff
@@ -41,6 +41,7 @@ dependencies {
     compile 'org.springframework.boot:spring-boot-starter-web'
     compile 'com.fasterxml.jackson.datatype:jackson-datatype-jsr310'
     compile "commons-io:commons-io:$commonsIOVersion"
+    compile group: 'org.apache.commons', name: 'commons-text', version: '1.4'
```
[the remark this answers is not in the document]
Comment (christophe-f) [reply build.gradle]: Will do

File: src/main/java/com/scotiabank/accelerator/initializer/controller/ComponentController.java [hunk 3]
```diff
@@ -29,16 +34,27 @@
...
         this.rootDir = checkNotNull(rootDir);
+        this.templateProcessor = templateProcessor;
     }
 
     @PostMapping("/api/project/generate")
-    public ResponseEntity<byte[]> userDownload(@Validated @RequestBody ProjectProperties component) {
-        byte[] content = projectCreationService.create(convertToProjectCreation(component));
+    public ResponseEntity<byte[]> userDownload(@Validated @RequestBody ProjectProperties component) throws InvalidTemplateException, IOException, URISyntaxException {
+        byte[] content;
+        if (component.getType().equals(ApplicationType.JAVA_SPRING_BOOT_2)) {
```
[the remark this answers is not in the document]
Comment (christophe-f) [reply src/main/java/com/scotiabank/accelerator/initializer/controller/ComponentController.java]: Yes the if statement is only until all the templates are migrated with the new engine. 

The switch in the TemplateProcessor will be replaced by the manifest once we have it. So we won't even need to add code when a template is added.

File: src/main/java/com/scotiabank/accelerator/initializer/controller/request/ProjectProperties.java [hunk 4]
```diff
@@ -21,11 +22,18 @@
...
     private String group;
-    
+
+    private String groupId;
+
     @NotNull
     private ApplicationType type;
 
+    @JsonIgnore
+    private String javaApplicationName;
```
[the remark this answers is not in the document]
Comment (christophe-f) [reply src/main/java/com/scotiabank/accelerator/initializer/controller/request/ProjectProperties.java]: I will move them in the TemplateProcessor. Ideally this is moved in the manifest files once we have them

File: src/main/java/com/scotiabank/accelerator/initializer/engine/TemplateProcessor.java [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+    /**
+     * Find the path of the template based on an application type
+     *
+     * @param applicationType {@link ApplicationType} of the project to be created
+     * @return URI of the template path
+     * @throws URISyntaxException       if the path is not a valid URI
+     * @throws NullPointerException     if the path is null
+     * @throws InvalidTemplateException if the {@link ApplicationType} is invalid
+     */
+    private URI getTemplatePath(ApplicationType applicationType) throws URISyntaxException, InvalidTemplateException {
```
[the remark this answers is not in the document]
Comment (christophe-f) [reply src/main/java/com/scotiabank/accelerator/initializer/engine/TemplateProcessor.java]: Agreed. Will add them back when the others templates are migrated.

File: src/main/java/com/scotiabank/accelerator/initializer/engine/TemplateProcessor.java [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+        return content;
+    }
+
+    /**
+     * Delete a directory and all its content recursively.
+     *
+     * @param path The path of the directory to delete
+     * @throws IOException if the path could not be deleted.
+     */
+    private void deleteDirectoryAndContent(String path) throws IOException {
```
[the remark this answers is not in the document]
Comment (christophe-f) [reply src/main/java/com/scotiabank/accelerator/initializer/engine/TemplateProcessor.java]: I found the root folder a bit complex and was not deleting the temp folder and files (at least on windows)

File: src/main/java/com/scotiabank/accelerator/initializer/engine/TemplateProcessor.java [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+
+        log.debug("Generating destination directory");
+        final Path destinationPath = createDestinationDirectory(component.getName());
+        log.trace("Destination template directory is {}", destinationPath);
+
+        // Candidate to be in a future manifest
+        String javaApplicationName = WordUtils.capitalizeFully(component.getName(), '.', ' ', '-');
+        javaApplicationName = javaApplicationName.replaceAll("[\\.\\-]", "");
+        component.setJavaApplicationName(javaApplicationName);
+        component.setJavaPackageName(component.getName().replaceAll("[\\.\\-]", ""));
```
[the remark this answers is not in the document]
Comment (christophe-f) [reply src/main/java/com/scotiabank/accelerator/initializer/engine/TemplateProcessor.java]: Yes I hope so. The manifest could have the input variables definitions including regex.

This one can have some extra brain storming

File: gradle/wrapper/gradle-wrapper.properties [hunk 8]
```diff
@@ -1,6 +1,6 @@
-#Fri Aug 10 16:56:18 EDT 2018
+#Mon Aug 20 13:16:16 EDT 2018
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
 zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-4.9-bin.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-4.9-all.zip
```
Comment (eddumelendez) [inline gradle/wrapper/gradle-wrapper.properties]: should we avoid this kind of changes?

File: src/main/java/com/scotiabank/accelerator/initializer/engine/TemplateProcessor.java [hunk 9]
```diff
@@ -0,0 +1,235 @@
...
+import com.scotiabank.accelerator.initializer.core.zip.ZipFile;
+import com.scotiabank.accelerator.initializer.model.ApplicationType;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.io.FileUtils;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.commons.text.WordUtils;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Component;
+
+import java.io.*;
```
Comment (eddumelendez) [inline src/main/java/com/scotiabank/accelerator/initializer/engine/TemplateProcessor.java]: avoid using *

Comment (christophe-f) [reply src/main/java/com/scotiabank/accelerator/initializer/engine/TemplateProcessor.java]: Was an IntelliJ optimization :) but will do if it's the project standards

File: gradle/wrapper/gradle-wrapper.properties (hunk 8 above)
Comment (christophe-f) [reply gradle/wrapper/gradle-wrapper.properties]: Good catch will update this.

File: src/main/java/com/scotiabank/accelerator/initializer/core/creator/impl/SpringBoot2ProjectCreatorImpl.java [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+package com.scotiabank.accelerator.initializer.core.creator.impl;
+
+import com.scotiabank.accelerator.initializer.core.creator.ProjectCreator;
+import com.scotiabank.accelerator.initializer.core.model.ProjectCreation;
+import com.scotiabank.accelerator.initializer.engine.TemplateProcessor;
+import com.scotiabank.accelerator.initializer.model.ApplicationType;
+import org.springframework.stereotype.Component;
+
+@Component
+class SpringBoot2ProjectCreatorImpl implements ProjectCreator<ProjectCreation> {
```
Comment (yshewchuk) [inline src/main/java/com/scotiabank/accelerator/initializer/core/creator/impl/SpringBoot2ProjectCreatorImpl.java]: This class is missing tests

File: src/main/java/com/scotiabank/accelerator/initializer/core/model/ProjectCreation.java [hunk 11]
```diff
@@ -33,13 +36,25 @@ public boolean isNodeApp() {
...
+    public String packageToPath(String content) {
+        return content.replaceAll("\\.", Matcher.quoteReplacement(File.separator));
+    }
+
     public String discoverySpringBootVersion() {
-        if(type == ApplicationType.JAVA_SPRING_BOOT_2) {
-            return "2.0.1.RELEASE";
-        }
-        return "1.5.7.RELEASE";
+        return "1.5.15.RELEASE";
```
Comment (yshewchuk) [inline src/main/java/com/scotiabank/accelerator/initializer/core/model/ProjectCreation.java]: Did you happen to verify that the project generated still works for this version of spring boot?

File: src/main/java/com/scotiabank/accelerator/initializer/core/model/ProjectCreation.java [hunk 12]
```diff
@@ -33,13 +36,25 @@ public boolean isNodeApp() {
     }
 
     public String resolvePackageName() {
-        return "com."+ getGroup().toLowerCase();
+        return "com." + getGroup().toLowerCase();
+    }
+
+    public String getJavaPackageName() {
```
Comment (yshewchuk) [inline src/main/java/com/scotiabank/accelerator/initializer/core/model/ProjectCreation.java]: Can we move this to a string utils class and rename it something like 'toLowerCase' and adjust the functionality to follow all package naming rules (ensure first character is letter or '_', all lowercase, ignore non-letters/digits)?
Also, this is lacking tests

File: src/main/java/com/scotiabank/accelerator/initializer/core/model/ProjectCreation.java [hunk 13]
```diff
@@ -33,13 +36,25 @@ public boolean isNodeApp() {
...
+        return getName()
+                .replaceAll("[\\.\\-]", "")
+                .toLowerCase();
     }
-    
+
+    public String getJavaApplicationName() {
```
Comment (yshewchuk) [inline src/main/java/com/scotiabank/accelerator/initializer/core/model/ProjectCreation.java]: Can we move this to a string utils class and rename it something like 'toPascalCase' and adjust the functionality (ensure first character is letter or '_', Pascal case, treat non-letters as word breaks, don't include characters that are not letters/digits)?
Also, this is lacking tests

File: src/main/java/com/scotiabank/accelerator/initializer/core/model/ProjectCreation.java [hunk 14]
```diff
@@ -33,13 +36,25 @@ public boolean isNodeApp() {
...
+        return WordUtils.capitalizeFully(getName(), '.', ' ', '-')
+                .replaceAll("[\\.\\-]", "");
+    }
+
+    public String packageToPath(String content) {
```
Comment (yshewchuk) [inline src/main/java/com/scotiabank/accelerator/initializer/core/model/ProjectCreation.java]: Can we move to a utility class and add tests?

File: src/test/java/com/scotiabank/accelerator/initializer/engine/TemplateProcessorTest.java [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+    private TemplateProcessor templateProcessor;
+
+    @Before
+    public void setUp() throws URISyntaxException {
+        sourceTemplateParentPath = Paths.get(getClass().getClassLoader().getResource("templates/projectCreation").toURI()).toString();
+        templateProcessor = new TemplateProcessor(resourceLoader, sourceTemplateParentPath);
+    }
+
+    @Test
+    public void processRelativePath_withDirectoryContainingVariables_shouldReturnProcessedRelativePath() {
```
Comment (yshewchuk) [inline src/test/java/com/scotiabank/accelerator/initializer/engine/TemplateProcessorTest.java]: I like these tests, but they're limited to only the processRelativePath method, is it possible to test create/process?

File: build.gradle [hunk 16]
```diff
@@ -37,18 +38,19 @@ sourceSets {
...
     compile "io.springfox:springfox-swagger2:$springfoxSwaggerVersion"
     compile "io.springfox:springfox-swagger-ui:$springfoxSwaggerVersion"
-    testCompile 'org.springframework.boot:spring-boot-starter-test'
-    testCompile 'com.jayway.jsonpath:json-path'
-    testCompile 'com.jayway.jsonpath:json-path-assert'
-    integrationTestCompile 'org.springframework.boot:spring-boot-starter-test'
+    testCompile "org.springframework.boot:spring-boot-starter-test"
+    testCompile "com.jayway.jsonpath:json-path"
+    testCompile "com.jayway.jsonpath:json-path-assert"
+    integrationTestCompile "org.springframework.boot:spring-boot-starter-test"
```
Comment (yshewchuk) [inline build.gradle]: Looks nice and clean!

**Comment by yshewchuk** [comment]
Overall looks much better - still some codacy issues to address and a number of things need to be tested

File: src/integrationTest/java/com/scotiabank/accelerator/initializer/controller/SpringBoot2ProjectTest.java [hunk 17]
```diff
@@ -36,7 +36,7 @@
     public void whenDownloadBodyIsOkThenExpect201() throws Exception {
         this.mvc.perform(post("/api/project/generate")
             .contentType(MediaType.APPLICATION_JSON_UTF8)
-            .content("{\"group\" : \"HOPPER\", \"type\" : \"JAVA_SPRING_BOOT_2\", \"name\": \"hopper-intake\"}")
+            .content("{\"group\" : \"HOPPER\", \"type\" : \"JAVA_SPRING_BOOT_2\", \"name\": \"hopper-intake\", \"groupId\": \"com.test\"}")
```
[the remark this answers is not in the document]
Comment (christophe-f) [reply src/integrationTest/java/com/scotiabank/accelerator/initializer/controller/SpringBoot2ProjectTest.java]: Yes, I missed that one.

File: src/main/java/com/scotiabank/accelerator/initializer/core/creator/impl/SpringBoot2ProjectCreatorImpl.java (hunk 10 above)
Comment (christophe-f) [reply src/main/java/com/scotiabank/accelerator/initializer/core/creator/impl/SpringBoot2ProjectCreatorImpl.java]: Will add more tests in the next commit, didn't want to postpone the review as it's been a while.

File: src/main/java/com/scotiabank/accelerator/initializer/core/model/ProjectCreation.java (hunk 11 above)
Comment (christophe-f) [reply src/main/java/com/scotiabank/accelerator/initializer/core/model/ProjectCreation.java]: yes no problem running the downloaded project

File: src/main/java/com/scotiabank/accelerator/initializer/core/model/ProjectCreation.java (hunk 12 above)
Comment (christophe-f) [reply src/main/java/com/scotiabank/accelerator/initializer/core/model/ProjectCreation.java]: Yes absolutely. I put it there as Mustache variables are named {{javaPackageName}} and expects a getter on it. I wanted to create a functional interface to fix this right after this PR. Let me know if you prefer to have it part of this one.

Comment (yshewchuk) [reply src/main/java/com/scotiabank/accelerator/initializer/core/model/ProjectCreation.java]: We can refactor it in following PR, but tests should be this PR

**Comment by christophe-f** [comment]
closes #33

**Pull request merged**